src/parsec_client/operations/psa_algorithm.md

```diff
@@ -0,0 +1,347 @@
+# PSA Algorithm
+
+The algorithm is used to select the specific cryptographic operation and to set a policy on a new
+key.
+
+*Notice:* not all possible attributes are currently supported by Parsec. Please see the [API
+coverage](../../api_coverage.md) for an overview of what Parsec currently supports.
+
+## Algorithm message
+
+| Field name | Type    | Description                                 |
+|------------|---------|---------------------------------------------|
+| `variant`  | `oneof` | Key type, see below for the possible values |
```

Comment:
I'm unsure whether we should be making this protobuf-specific, since theoretically that could be replaced with anything.

Reply:
I was thinking about that as well! Was unsure if, in the book, the operation documentation should be common to all body types or not.

If it is, I should modify this PR to make it totally generic: do not assume any types or any name. Basically take the PSA Crypto API specification and translate it into a generic Parsec version, language agnostic. Then we would have to document how from this generic model, the different things are implemented in all body type languages like protobuf and others.

If we choose to make it body type language specific, then we would have to add one more level on the book below the "Operations" section for each body type language. That would mean some repetition in the documentation.

Reply:
Not sure if we can decide that for ourselves!

@paulhowardarm - any thoughts on this? Should these docs (the ones in the operations folder) be specific to protobuf?

Reply:
The original idea was for the operation directory to be neutral, and for the "Contract" section at the bottom to provide the links to interface definitions such as protobuf. So the expected developer flow was that you'd start off with the markdown document to get the general overview of the operation, and a neutral description of the ins/outs, you'd then jump the hyperlink to the (say) protobuf file to see the contract and any additional docs on a per-field or per-parameter basis.

Now there are problems with this original view. The first problem is that documentation looks far better in markdown than it does in protobuf where it's just unformatted text. I thought that we could solve this by auto-generating MD docs from the protobuf files, but then you need a build step to generate the docs and you can't just read it in GitHub. I think it's important that our documentation should be readable from the source without requiring a build, so auto-generation doesn't work. The second problem is that you would need to replicate all that stuff when a new serialisation format comes along (if indeed that ever happens - there's no immediate prospect of that).

If all our API doc is to be static MD with no auto-generation step, then I think the gold standard would be for all of the doc to be format-neutral, such that you can read it and then jump the hyperlink to the protobuf (or whatever) file, and the content in that file should be self-evident from what you've been reading, without any significant extra doc in that file. This doesn't have to mean that the doc was protobuf-specific, but it should have been descriptive enough to make the protobuf understandable by itself.

The question is whether that can be achieved in practice. If you look at what's there today, some of our existing content follows that ideal. For example, the **PsaCreateKey** MD file has a "**Parameters**" section with `key_name` and `key_attributes`. And when you jump to the protobuf file, you see a protobuf `Operation` message with fields `key_name` and `attributes` and the precise data types. In my head, that's very clear and self-evident, and yet we've managed to keep the MD document format-neutral.

So the question is: can we do that everywhere? Can we explain key attributes in a way that explains what it means, and what all the options are, but then link to the protobuf file in such a way that it simply confirms what we've already written, or will contain content that exactly matches what the reader will expect to see?

If we can do that, then I think it's the winning solution.

If we can't, then probably the next best thing is to change the hyperlinks so that "Protobuf" doesn't go straight to the protobuf file itself, but instead goes to an interim MD file that explains whatever part of the contract absolutely requires a protobuf-specific fragment of doc, and then have that interim MD file have another hyperlink that goes to the protobuf message file itself. A level of indirection, in other words. But we only do it in a few well-judged places, rather than structure the entire book in this way.

Does that make sense?

Reply:
Thanks for the detailed explanation! That makes a lot of sense. Having format-neutral documentation is the best thing to do to avoid any kind of duplication and to have more readable documentation as well.
I think for most operations/structures that we describe in a format-neutral way from the PSA Crypto API, the translation and understanding of the Protobuf format will be implicit.

Will modify this PR then, to put it back to be formal-neutral!